client/kafka/kafka.go

```diff
@@ -5,45 +5,73 @@ import (
 	"errors"
 	"fmt"
 
-	"github.com/Shopify/sarama"
 	"github.com/beatlabs/patron/correlation"
 	"github.com/beatlabs/patron/encoding"
-	"github.com/beatlabs/patron/trace"
+
+	"github.com/Shopify/sarama"
 	"github.com/opentracing/opentracing-go"
-	"github.com/opentracing/opentracing-go/ext"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
 const (
-	producerComponent     = "kafka-async-producer"
-	messageCreationErrors = "creation-errors"
-	messageSendErrors     = "send-errors"
-	messageSent           = "sent"
+	asyncProducerComponent = "kafka-async-producer"
+	syncProducerComponent  = "kafka-sync-producer"
+	messageCreationErrors  = "creation-errors"
+	messageSendErrors      = "send-errors"
+	messageSent            = "sent"
 )
 
-var messageStatus *prometheus.CounterVec
+var (
+	asyncMessageStatus *prometheus.CounterVec
+	syncMessageStatus  *prometheus.CounterVec
+)
+
+// Producer interface for Kafka.
+type Producer interface {
+	Send(ctx context.Context, msg *Message) error
+	Close() error
+}
 
-func messageStatusCountInc(status, topic string) {
-	messageStatus.WithLabelValues(status, topic).Inc()
+type baseProducer struct {
+	cfg           *sarama.Config
+	prodClient    sarama.Client
+	tag           opentracing.Tag
+	enc           encoding.EncodeFunc
+	contentType   string
+	messageStatus *prometheus.CounterVec
+}
+
+var (
+	_ Producer = &AsyncProducer{}
+	_ Producer = &SyncProducer{}
+)
+
+// Message abstraction of a Kafka message.
+type Message struct {
+	topic string
+	body  interface{}
+	key   *string
 }
 
 func init() {
-	messageStatus = prometheus.NewCounterVec(
+	asyncMessageStatus = prometheus.NewCounterVec(
```

Comment:
we could have just one counter with an extra label "type": "async" or "sync"

Reply:
@mantzas that means we break with the existing counter `kafka_async_producer` because the new one would be called `kafka_producer`. Are we OK with that?

Reply:
You are right. Since we move towards 1.0 we can break it. After 1.0 we will be more restricted in breaking things.

Reply:
@mantzas I have changed it to use a `type` label now